lib/auth/userpreferences/userpreferencesv1/service.go

```diff
@@ -21,42 +21,72 @@ import (
 	"context"
 
 	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
 	"google.golang.org/protobuf/types/known/emptypb"
 
 	userpreferences "github.com/gravitational/teleport/api/gen/proto/go/userpreferences/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/authz"
 	"github.com/gravitational/teleport/lib/services"
 )
 
 // ServiceConfig holds configuration options for the user preferences service.
 type ServiceConfig struct {
-	Backend services.UserPreferences
+	Backend    services.UserPreferences
+	Authorizer authz.Authorizer
+	Logger     *logrus.Entry
 }
 
 // Service implements the teleport.userpreferences.v1.UserPreferencesService RPC service.
 type Service struct {
 	userpreferences.UnimplementedUserPreferencesServiceServer
 
-	backend services.UserPreferences
+	backend    services.UserPreferences
+	authorizer authz.Authorizer
+	log        *logrus.Entry
 }
 
 // NewService returns a new user preferences gRPC service.
 func NewService(cfg *ServiceConfig) (*Service, error) {
 	switch {
 	case cfg.Backend == nil:
 		return nil, trace.BadParameter("backend is required")
+	case cfg.Authorizer == nil:
+		return nil, trace.BadParameter("authorizer is required")
+	case cfg.Logger == nil:
+		cfg.Logger = logrus.WithField(trace.Component, "userpreferences.service")
 	}
 
 	return &Service{
-		backend: cfg.Backend,
+		backend:    cfg.Backend,
+		authorizer: cfg.Authorizer,
 	}, nil
 }
 
 // GetUserPreferences returns the user preferences for a given user.
 func (a *Service) GetUserPreferences(ctx context.Context, req *userpreferences.GetUserPreferencesRequest) (*userpreferences.GetUserPreferencesResponse, error) {
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if authCtx.User.GetName() != req.GetUsername() {
+		return nil, trace.AccessDenied("user %q is not allowed to access preferences for user %q", authCtx.User.GetName(), req.Username)
+	}
+
 	return a.backend.GetUserPreferences(ctx, req)
 }
 
 // UpsertUserPreferences creates or updates user preferences for a given username.
 func (a *Service) UpsertUserPreferences(ctx context.Context, req *userpreferences.UpsertUserPreferencesRequest) (*emptypb.Empty, error) {
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if authCtx.User.GetName() != req.GetUsername() {
+		return nil, trace.AccessDenied("user %q is not allowed to access preferences for user %q", authCtx.User.GetName(), req.Username)
+	}
```

Comment:
I think we want to mimic the `currentUserAction` method used in `auth_with_roles`:

```
// AuthorizeCurrentUser will ensure that the user has access to the given resource based on ownership, or based
// on the given verbs for the given kind.
func AuthorizeCurrentUser(ctx context.Context, log logrus.FieldLogger, authorizer Authorizer, quiet bool, owner string, kind string, verbs ...string) (*Context, error) {
	authCtx, err := authorizer.Authorize(ctx)
	if err != nil {
		return nil, ConvertAuthorizerError(ctx, log, err)
	}

	if authCtx.User.GetName() == owner {
		return authCtx, nil
	}

	ruleCtx := &services.Context{
		User: authCtx.User,
	}

	return authorizeContextWithVerbs(ctx, log, authCtx, quiet, ruleCtx, kind, verbs...)
}
```

```suggestion
	authCtx, err := authz.AuthorizeCurrentUser(ctx, a.log, a.authorizer, true, req.GetUsername(), types.KindUser, types.VerbUpdate)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	if authCtx.User.GetName() != req.GetUsername() {
		return nil, trace.AccessDenied("user %q is not allowed to access preferences for user %q", authCtx.User.GetName(), req.Username)
	}
```

Reply:
Or, given the comment above to remove the username field:

```
func AuthorizeLocalUser(ctx context.Context, log logrus.FieldLogger, authorizer Authorizer, quiet bool) (*Context, error) {
	authCtx, err := authorizer.Authorize(ctx)
	if err != nil {
		return nil, ConvertAuthorizerError(ctx, log, err)
	}

	if _, ok := authCtx.UnmappedIdentity.(LocalUser); !ok {
		return nil, trace.AccessDenied("action not permitted for non-local user")
	}

	return authCtx, nil
}
```

Reply:
I think it should be fine now.